ethereum/contracts/Bridge.sol

```diff
@@ -223,42 +259,29 @@ contract Bridge is AccessControl {
     // ************    ADMIN-LEVEL ACCESS METHODS   *************
 
 
-    function inflate(uint256 amount)
+    function mint(uint256 amount)
```

Comment:
suggest a different implementation here (cap is enforced on minting rather than aligned with it):

 function mint(uint256 amount)
        public
        onlyOwner
    {
       effectiveAmountAccrued = effectiveAmountAccrued.add(amount);
       require(cap >= effectiveAmountAccrued, "minting exceeds transfer cap");
       inflationAmount = inflationAmount.add(amount);
       emit Inflate(amount);
       token.mint(address(this), amount);
    }

Reply:
The `cap` state variable was originally designed to limit value of the `effectiveAmountAccrued` state variable (the amount transferred accros the bridge), so these two state variable were tight up on logical level.

When the whole token supply is affected by inflation/deflation (= minting/burning), then `effectiveAmountAccrued` amount of tokens users transferred across the bridge must be also inflated/deflated accordingly to preserve their "real world" value. Implication is, that `cap` value must be also inflated/deflated by the same amount, since it is logically tight up with `effectiveAmountAccrued`.

I actually wanted to intorduce dedicated cap/limit for minting in this PR, but I simply forgot to do that. This limit would be controlling `inflationAmount` value, which would need to be signed initeg (`int256`), and so the SafeMath library can not be used - overflow would need to be controlled mabually.

Reply:
I think there's a cleaner way of thinking about this. 
1. The only variable that really matters is the token.balanceOf(address(this)), which tracks the total supply on the native chain (it's fine to have another state variable to represent this). 
2. The minting and burning are used to sync the token supply in the contract with the supply on the native-chain.
3. The cap is used to limit the supply of tokens on the native-chain. 
4. There's no need to have a separate withdrawFees function as all of the transfers go to the native-chain and any fees are withheld there. 
5. There can be a single withdrawFunds function that moves all the tokens in the contract, and is expected to only be called when we move to a different bridge contract.

Reply:
It can be more-or-less symmetric on the other side with the important difference that the fees for both sides are not stored in the contract. It's sensible to do it this way as the tx costs are so much lower than on the ethereum chain.

Reply:
oh, I realised there might also be a discrepancy between the token.balanceOf(address(this)) and the amount tracked/transferred if someone accidentally sends tokens to the contract address. We could have another function to refund those.

Reply:
The `cap` state variable was originally designed to limit how much users can ransfer transferred accros the bridge = in `swap(...)` call, not to limit how much can be minted.

The difference is, that when the whole token supply is affected by inflation/deflation (= minting/burning), then `effectiveAmountAccrued` amount of tokens users transferred across the bridge must be also inflated/deflated accordingly to **preserve** their "real world" value (= how much of real world assets (e.g. bread, milk) can be bought for tokens). Implication is, that `cap` value must be also inflated/deflated by the same amount, since it is logically tight up with `effectiveAmountAccrued` (= limiting how much users can transfer acros), **not** with minting(= how much can be minted (or burned - see bellow).
Burning (the inverse action) is actually nice reason why standing `cap` state variable does not fit quite well, since it can not limit burning process (e.g. limit how much can be burned), the burn limit would be implicitelly the whole `effectiveAmountAccrued` amount.

I wanted to introduce 2 limit state variables for minting/burning, which would be logically tihgt up with `inflationAmount` state variable.However I decided not to do that, since it

Reply:
There is function for refunding excess tokens (= discrepancy `between token.balanceOf(address(this))` and `effectiveAmountAccrued + feesAccrued)`): `withdrawExcessTokens(address payable targetAddress)`

Reply:
About point 5 aboue: There is withdraw "all funds" function already present - the `deleteContract(...)`.and it is by- design

Reply:
There is also dedicated function for withdrawing fees accrued in the cotract: `withdrawFees(address targetAddress)`

Reply:
I get what you're saying Peter, but I think we mostly care about controlling the number of tokens on main-net rather than the flow. Secondly, it would be nice to retain the flexibility of keeping this contract around and withdrawing only some of the funds. Something like a `withdraw(address payable targetAddress, uint256 amount)`

Reply:
Mathematically speaking, the current implementation of the contract offers mathematically correct handling of all the folloowing aspects of the bridge:

Reply:
Also because the fees are going to be taken on the other side of the contract, there's no need to keep track of the fees.

Reply:
But, in negeral, and in practice, fee setting on each side(contract) can be, and should be, different.This is due to since that fee user pays is for us payun

Reply:
Another thing about withdrawal: Doing wild withdrawals from the contract exposes contract to potential "double-spend" lile situation where certain amount of tokens will be present in circulating supply on both side (etherem and our mainnet.

Such "withdrawal" can be achieved now by correct way - doing official swap (on cosmos side) whi would be normally transferred to the ethereum side using normal swap process = swapped amount will "disappear" (get locked) on cosmos side, and will apper on Ethereum side on the destination address specified in the swap.

Reply:
Resolved in following commits - please see [my comment](https://github.com/fetchai/fetch-ethereum-bridge-v1/pull/2#issuecomment-770809906) sumarising the changes reflecting requirements which eraised during by this discussion.